RIOT-OS/RIOT
# makefiles: cleanup DEFAULT_MODULE, introduce DEFAULT_MODULE_DELAYED

### Contribution description

This PR attempts to cleanup, limits and documents the use of `DEFAULT_MODULE` and `DISABLE_MODULE`. It is the result of findings in #13651, as well as as findings and discussion while reviewing #13349.

**(a)** This PR includes `DEFAULT_MODULE`s only once to avoid undefined behavior where `DEFAULT_MODULES` added during the first dependency resolution iteration would be converted to `USEMODULE` and others wouldn't. If a `BOARD` or `CPU` needs a `DEFAULT_MODULE` added before dependency resolution it may add it to a `$(RIOTCPU)/$(CPU)/defaultmodules.inc.mk` file. `DEFAULT_MODULE`s added here can have dependencies, as they are resolved before dependency resolution (although there is currently no use case for this). This also means they can only be disabled in a declarative way (no `ifneq`) by an `APPLICATION`, `CPU` or `BOARD`.

~~This means that current handling of `stdio_%` is not allowed, but we keep it as is until another solution is proposed.~~ Now removed

The case where a `DEFAULT_MODULE` needs conditional disabling will be unsupported, and with no intentions to support through this mechanism. 

**(b)** Any `DEFAULT_MODULE` added in a `Makefile.dep` will only be added at the end of dependency resolution process. They MAY not have any dependencies as they will never be converted during resolution. This Modules MAY be DISABLED at any time, since they are only resolved once at the end of dependency resolution.

This PR also introduces a new `DEFAULT_MODULE_DELAYED`. This variable allows making a difference between the two cases mentioned above. `DEFAULT_MODULE` is used for **(a)** and the new `DEFAULT_MODULE_DELAYED` for **(b)**.

Adding the new variable allows to document the expected properties of that module. It would make it easier to add static checks verifying allowed behaviors easier. But if deemed unnecessary the last two commits adding and using `DEFAULT_MODULE_DELAYED ` could be removed. I could also split that change as well as the `auto_init`/`periph_init` change.

Here is a debrief of the changes commit by commit.

- d2d50eba71ef35a0f6b69590b6588252c5d00cbc
 converts `DEFAULT_MODULE` to `USEMODULE` only before and after dependency resolution. This avoid undefined behaviour where depending on order on inclusion some modules would be added as `USEMODULE` during dependency resolution and others wouldn't.
- ~~~4fd41603af60b2e28c6c2378b57b70ec2430b5c7 adds an exception for `stdio_`~~
- db858fa468df7ef26deb1be1293583471c4a43e3 enables adding `DEFAULT_MODULE` at `BOARD` and `CPU` level, this wants to get ready for when `DEFAULT_MODULE` addition is limited
- 991cecf  moves `auto_init` and `periph_init` module disabling. Removes application wildcard disable.
- 36f61d217ab6fc82d36b86d6474ca9a43ffaeb0b: adds a mew `DEFAULT_MODULE_DELAYED` variable and all required modifications to support it.
- 0f356beb61f97da244fad560f912e9f4182bd172 replaces all `DEFAULT_MODULE` not in `defaultmodules.in.mk` by `DEFAULT_MODULE_DELAYED`


TODO (as follow-ups):

- [ ] entries in `build-system-basics`
- [ ] add static checks
- [x] remove `stdio_%` special handling

~~To not conflict with #13349 this PR is based on it, although the changes and objectives are orthogonal.~~

Therefore this PR is reviewed better commit by commit, to not mix the changes in #13349. ~~I'll rebase as soon as possible to remove those.~~ rebased

### Testing procedure

- green murdock

- `save_all_dependencies` should be ran but there will be many differences because of variable change. I would suggest removing 1ea4fc4 to reduce noise. There should not be any difference with regard to `USEMODULE`s, differences should show up only in `DEFAULT_MODULE` and `DEFAULT_MODULE_DELAYED`

- feedback on naming for `DEFAULT_MODULE_DELAYED`

- all `periph_init_%` and `auto_init_%` modules are still present and can be removed:

`BOARD=nucleo-l152re make -C tests/xtimer_usleep/ info-debug-variable-USEMODULE --no-print-directory`

```
auto_init_xtimer board boards_common_nucleo core core_init core_msg core_panic cortexm_common cortexm_common_periph cpu div isrpipe newlib newlib_nano newlib_syscalls_default periph periph_common periph_gpio periph_init_gpio periph_init_pm periph_init_timer periph_init_uart periph_pm periph_timer periph_uart pm_layered stdin stdio_uart stdio_uart_rx stm32_common stm32_common_periph sys test_utils_interactive_sync tsrb xtimer
```

 `DISABLE_MODULE="periph_init_gpio auto_init" BOARD=nucleo-l152re make -C tests/xtimer_usleep/ info-debug-variable-USEMODULE --no-print-directory`
```
board boards_common_nucleo core core_init core_msg core_panic cortexm_common cortexm_common_periph cpu div isrpipe newlib newlib_nano newlib_syscalls_default periph periph_common periph_gpio periph_init_pm periph_init_timer periph_init_uart periph_pm periph_timer periph_uart pm_layered stdin stdio_uart stdio_uart_rx stm32_common stm32_common_periph sys test_utils_interactive_sync tsrb xtimer
```

### Issues/PRs references

Addresses part of #13651, will be finished when TODO list is completed.
